java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -164,42 +185,137 @@ sealed class GlobalFunction(val name: String) {
+    }
   }
 
   /** Find the maximum of a [Sequence]. */
   object Max : GlobalFunction("max") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Comparable<Comparable<*>>>(args[0]).max()
+
+    override fun inferredType(
+      expr: Expression<*>,
+      evaluator: TypeEvaluator,
+      args: List<InferredType>
+    ): InferredType {
+      return checkInnerSequenceType("max", args, expr, evaluator)
+    }
   }
 
   /** Find the minimum of a [Sequence]. */
   object Min : GlobalFunction("min") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Comparable<Comparable<*>>>(args[0]).min()
+
+    override fun inferredType(
+      expr: Expression<*>,
+      evaluator: TypeEvaluator,
+      args: List<InferredType>
+    ): InferredType {
+      return checkInnerSequenceType("min", args, expr, evaluator)
+    }
   }
 
   /** Find the average of a [Sequence]. */
   object Average : GlobalFunction("average") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Int>(args[0]).average()
+
+    override fun inferredType(
+      expr: Expression<*>,
+      evaluator: TypeEvaluator,
+      args: List<InferredType>
+    ): InferredType {
+      checkInnerSequenceType("average", args, expr, evaluator)
+      return InferredType.Primitive.DoubleType
+    }
   }
 
   /** Find the average of a [Sequence]. */
   object Sum : GlobalFunction("sum") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Int>(args[0]).sum()
+
+    override fun inferredType(
+      expr: Expression<*>,
+      evaluator: TypeEvaluator,
+      args: List<InferredType>
+    ): InferredType {
+      return checkInnerSequenceType("sum", args, expr, evaluator)
+    }
   }
 
   /** Count the number of elements in a [Sequence]. */
   object Count : GlobalFunction("count") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Any>(args[0]).count()
+
+    override fun inferredType(
+      expr: Expression<*>,
+      evaluator: TypeEvaluator,
+      args: List<InferredType>
+    ): InferredType {
+      checkInnerSequenceType("count", args, expr, evaluator) { true }
+      return InferredType.Primitive.IntType
+    }
   }
 
   /** Return the first item of a [Sequence]. */
   object First : GlobalFunction("first") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Any>(args[0]).firstOrNull()
+
+    override fun inferredType(
+      expr: Expression<*>,
+      evaluator: TypeEvaluator,
+      args: List<InferredType>
+    ): InferredType {
+      checkInnerSequenceType("first", args, expr, evaluator) { true }
+      return InferredType.UnionType(
+        InferredType.Primitive.NullType,
+        when (val it = args[0]) {
+          is InferredType.SeqType -> it.type
+          else -> it
```

Comment:
I believe in `checkInnerSequenceType` above you enforce that the only argument is a sequence, so here you can rely on it. So instead you co just do:
```
return InferredType.UnionType(
  InferredType.Primitive.NullType,
  (args[0] as SeqType).type
)
```